mlir/lib/ExecutionEngine/execution_engine.cpp

```diff
@@ -0,0 +1,299 @@
+      LLVM_DEBUG(llvm::dbgs() << "No object for " << m->getModuleIdentifier()
+                              << " in cache. Compiling.\n");
+      return nullptr;
+    }
+    LLVM_DEBUG(llvm::dbgs() << "Object for " << m->getModuleIdentifier()
+                            << " loaded from cache.\n");
+    return llvm::MemoryBuffer::getMemBuffer(i->second->getMemBufferRef());
+  }
+
+  /// Dump cached object to output file `filename`.
+  void dumpToObjectFile(llvm::StringRef outputFilename) {
+    // Set up the output file.
+    std::string errorMessage;
+    auto file = mlir::openOutputFile(outputFilename, &errorMessage);
+    if (!file) {
+      llvm::errs() << errorMessage << "\n";
+      return;
+    }
+
+    // Dump the object generated for a single module to the output file.
+    assert(cachedObjects.size() == 1 && "Expected only one object entry.");
+    auto &cachedObject = cachedObjects.begin()->second;
+    file->os() << cachedObject->getBuffer();
+    file->keep();
+  }
+
+private:
+  llvm::StringMap<std::unique_ptr<llvm::MemoryBuffer>> cachedObjects;
+};
+
+/// Wrap a string into an llvm::StringError.
+static llvm::Error makeStringError(const llvm::Twine &message) {
+  return llvm::make_error<llvm::StringError>(message.str(),
+                                             llvm::inconvertibleErrorCode());
+}
+
+// Setup LLVM target triple from the current machine.
+static bool setupTargetTriple(llvm::Module *llvmModule) {
+  // Setup the machine properties from the current architecture.
+  auto targetTriple = llvm::sys::getDefaultTargetTriple();
+  std::string errorMessage;
+  const auto *target =
+      llvm::TargetRegistry::lookupTarget(targetTriple, errorMessage);
+  if (!target) {
+    llvm::errs() << "NO target: " << errorMessage << "\n";
+    return true;
+  }
+
+  std::string cpu(llvm::sys::getHostCPUName());
+  llvm::SubtargetFeatures features;
+  llvm::StringMap<bool> hostFeatures;
+
+  if (llvm::sys::getHostCPUFeatures(hostFeatures))
+    for (auto &f : hostFeatures)
+      features.AddFeature(f.first(), f.second);
+
+  std::unique_ptr<llvm::TargetMachine> machine(target->createTargetMachine(
+      targetTriple, cpu, features.getString(), {}, {}));
+  if (!machine) {
+    llvm::errs() << "Unable to create target machine\n";
+    return true;
+  }
+  llvmModule->setDataLayout(machine->createDataLayout());
+  llvmModule->setTargetTriple(targetTriple);
+  return false;
+}
+
+imex::ExecutionEngine::ExecutionEngine(ExecutionEngineOptions options)
+    : cache(options.enableObjectCache ? new SimpleObjectCache() : nullptr),
+      gdbListener(options.enableGDBNotificationListener
+                      ? llvm::JITEventListener::createGDBRegistrationListener()
+                      : nullptr),
+      perfListener(nullptr) {
+  if (options.enablePerfNotificationListener) {
+    if (auto *listener = llvm::JITEventListener::createPerfJITEventListener())
+      perfListener = listener;
+    else if (auto *listener =
```

Comment:
And if createIntelJITEventListener returns nullptr too?

Reply:
This was directly copypasted from mlir examples. It both functions fail, execution engine should still work, just without debugging support.